### PR TITLE
bugtool: get bpffs mountpoint from /proc/mounts

### LINKED ISSUE #13218: bugtool: Incorrect BPF filesystem path
In bugtool, the path to the BPF filesystem is hardcoded to `/sys/fs/bpf`:
https://github.com/cilium/cilium/blob/f55458f9bfd412b7334b535c43bab593910c6b31/bugtool/cmd/configuration.go#L67-L70

In some environments, however, the BPF filesystem is mounted elsewhere. Our bugtool should retrieve the correct path automatically. Without this we may be missing some map contents (e.g., CT maps) in some environments.

### PR DESCRIPTION
**Pull request by jibi**
Rather then hardcoding the /sys/fs/bpf value in bugtool, parse
/proc/mounts to determine the correct mountpoint for the BPF filesystem.

Fixes: #13218

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

### CODE REVIEW

File: bugtool/cmd/configuration.go [hunk 1]
```diff
@@ -39,6 +41,26 @@ func setupDefaultConfig(path string, k8sPods []string, confDir, cmdDir string) (
 	return &c, save(&c, path)
 }
 
+func bpffsMountpoint() string {
+	mounts, err := os.Open("/proc/mounts")
+	if err != nil {
+		return ""
+
```
Comment (qmonnet) [inline bugtool/cmd/configuration.go]: Nit: blank line

**Review by qmonnet — APPROVED** [review]
Looks great, thank you!

This won't be “tested” by the CI, I suppose you validated the change locally to make sure we still dump the maps as intended?

**Comment by qmonnet** [comment]
I'm just thinking: Although unlikely, we could have several bpffs mounted at different points. Would it be worth to try and list the maps under each of them in that case?

**Comment by jibi** [comment]
> Looks great, thank you!
> 
> This won't be “tested” by the CI, I suppose you validated the change locally to make sure we still dump the maps as intended?

Yes, I tested it both manually in the dev VMs as well as by crashing some e2e tests and looking at its dump in `/tmp`

**Comment by jibi** [comment]
> I'm just thinking: Although unlikely, we could have several bpffs mounted at different points. Would it be worth to try and list the maps under each of them in that case?

I wasn't aware of this possibility but yes, I agree that we should try list maps under each of the mounted bpffs :+1:

**Review by tklauser — APPROVED** [review]

**Comment by qmonnet** [comment]
> I wasn't aware of this possibility but yes, I agree that we should try list maps under each of the mounted bpffs +1

You can mount multiple ones, but I don't think this is something Cilium would ever do. Just thinking that if for any reason users have mounted another bpffs, we would risk missing the correct one when dumping the maps.

**Review by pchaigno — APPROVED** [review]

File: bugtool/cmd/configuration.go [hunk 2]
```diff
@@ -39,6 +41,27 @@ func setupDefaultConfig(path string, k8sPods []string, confDir, cmdDir string) (
 	return &c, save(&c, path)
 }
 
+func bpffsMountpoint() []string {
```
Comment (vadorovsky) [inline bugtool/cmd/configuration.go]: We already have a separate module for listing mounts from `/proc/self/mountinfo` (which contains more details than `/proc/mounts`):

https://github.com/cilium/cilium/blob/master/pkg/mountinfo/mountinfo.go#L125-L135

The example of using it, also for checking BPFFS mount, is here:

https://github.com/cilium/cilium/blob/master/pkg/bpf/bpffs_linux.go#L69-L81

**Review by vadorovsky — CHANGES_REQUESTED** [review]
Thanks for the PR!

We already have a separate package for listing mounts (`pkg/mountinfo`), I think it would be better to use it here.

**Comment by mrostecki** [comment]
@qmonnet @jibi 

> You can mount multiple ones, but I don't think this is something Cilium would ever do. Just thinking that if for any reason users have mounted another bpffs, we would risk missing the correct one when dumping the maps.

In cilium-agent we even return an explicit error if there are multiple BPFFS mounts:

https://github.com/cilium/cilium/blob/master/pkg/bpf/bpffs_linux.go#L278-L284

File: bugtool/cmd/configuration.go (hunk 2 above)
Comment (jibi) [reply bugtool/cmd/configuration.go]: Thanks @mrostecki, I'll take a look at that package now

**Comment by jibi** [comment]
> @qmonnet @jibi
> 
> > You can mount multiple ones, but I don't think this is something Cilium would ever do. Just thinking that if for any reason users have mounted another bpffs, we would risk missing the correct one when dumping the maps.
> 
> In cilium-agent we even return an explicit error if there are multiple BPFFS mounts:
> 
> https://github.com/cilium/cilium/blob/master/pkg/bpf/bpffs_linux.go#L278-L284

It looks like we have a case of multiple bpf fs even in the dev VM:

```sh
vagrant@k8s1:/sys/fs/bpf$ cat /proc/mounts | grep bpf
bpffs /sys/fs/bpf bpf rw,relatime 0 0
bpffs /sys/fs/bpf/xdp bpf rw,relatime 0 0
bpffs /sys/fs/bpf/ip bpf rw,relatime 0 0
bpffs /sys/fs/bpf/sk bpf rw,relatime 0 0
bpffs /sys/fs/bpf/sa bpf rw,relatime 0 0
```
and I'm slightly confused as `tc` (which is the only one we are interested in) is just a subdirectory of the main mountpoint (`/sys/fs/bpf`) while `xdp`, `ip`, `sk` and `sa` are all additional bpf filesystems :thinking:

**Comment by qmonnet** [comment]
Hmm I wonder why this does not seem to be picked up by the check in `bpffs_linux.go` :thinking:

**Comment by mrostecki** [comment]
@jibi ´/sys/fs/bpf/xdp`, `/sys/fs/bpf/ip`, `/sys/fs/bpf/sk` and `/sys/fs/bpf/sa` are sub mounts of `/sys/fs/bpf` (it means `/sys/fs/bpf´ is their parent mount`). So it's fine.

We want to avoid multiple BPFFS mounts which have `/` as their root. As we do it here: https://github.com/cilium/cilium/blob/master/pkg/bpf/bpffs_linux.go#L162

**Comment by mrostecki** [comment]
@qmonnet The last link I pasted is the reason why cilium-agent doesn't fail because of `/sys/fs/bpf´ subdirectories. ;)

**Comment by jibi** [comment]
Thank you both @qmonnet and @mrostecki, I just learnt a bunch of new things :) 

In this case I think i will:
* assume there's only a single bpffs mounted
* consider it to be the first one returned by `mountinfo.GetMountInfo()` which has both the `bpf` type and and the `/` root

File: bugtool/cmd/configuration.go [hunk 3]
```diff
@@ -39,6 +40,21 @@ func setupDefaultConfig(path string, k8sPods []string, confDir, cmdDir string) (
...
+	mountInfos, err := mountinfo.GetMountInfo()
+	if err != nil {
+		return ""
+	}
+
+	for _, mountInfo := range mountInfos {
+		if mountInfo.FilesystemType == "bpf" && mountInfo.Root == "/" {
+			return mountInfo.MountPoint
+		}
+	}
```
Comment (qmonnet) [inline bugtool/cmd/configuration.go:65]: Can you please add a comment above this block to explain 1) why only list one bpffs and 2) why the additional check on the `"/"` root? Might be helpful to remember the reasoning in the future.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks good to me, please just add a comment so we can remember the logics here.

**Review by qmonnet — APPROVED** [review]
Looks good to me, thanks!

**Review by vadorovsky — APPROVED** [review]
Looks good. Thanks!

**Pull request merged**